### DISCUSSION TITLE
Page reloads when interacting with system modal dialogs from the browser

### DISCUSSION THREAD

**Discussion by thetableman** [zauberzeug/nicegui#2286] [Q&A]
### Question

Consider the below code that calls the javascript function `window.print()`.

```py
from nicegui import ui

async def print_dialog():
    ui.run_javascript('window.print();')

ui.button('Print', on_click=print_dialog)

ui.run()
```

The javascript executes correctly but if the `window.print()` dialog is kept open for more than a few seconds, to adjust print settings or to navigate to a folder for saving a pdf, the page reloads afterwards.

Because `window.print()` doesn't return any response you can't await it as this results in a `TimeoutError: JavaScript did not respond in time` error.

Does anyone have any suggestions to execute this javascript without the page reloading afterwards?

**Comment by rodja** [comment]
I could not reproduce the issue with Firefox, Chrome or Safari on Mac. What Browser/OS are you using?

**Reply by thetableman** [reply]
Win 10 & 11, Firefox & Chrome.

How long did you wait?

**Reply by rodja** [reply]
After some more testing, I managed to reproduce it:

![print-forces-reload](https://github.com/zauberzeug/nicegui/assets/131391/84240024-98d4-46c3-90a3-bfca6411bfaf)

```py
@ui.page('/')
async def index():
    ui.label(f'start time {datetime.now().strftime("%H:%M:%S %f")}')
    time_label = ui.label('current time')
    ui.timer(0.1, lambda: time_label.set_text(datetime.now().strftime('%H:%M:%S %f')))

    async def print_dialog():
        ui.run_javascript('window.print();')

ui.run(reconnect_timeout=3)
```

As you can see in the GIF, the start time changes, so the page did reload. Oddly the same code sometimes ran perfectly in the background (for example when opening the debug console): 

![print-allows-bg](https://github.com/zauberzeug/nicegui/assets/131391/6b2bb18f-3006-4af6-99b2-2c502bab7da1)

Does anybody know when js / socket communication is still allowed to run in the background and how the browser decides to pause the tab?

**Reply by retsyo** [reply]
I met similar page reload. In my case, `ui.run_javascript` exports current page to PDF like https://github.com/zauberzeug/nicegui/discussions/2025 says. No dialog shows here, but it costs some seconds to finish the export processing, 

The reload happens frequently if I use mobile phone to visit page.

However, if I visit page using webbrowsers on Windows 10 64bits, the reload does not appear so often.

Since I delete current user info before `ui.run_javascript`, the reload actually send current user to log page, and the exported PDF is blank.

**Comment by rodja** [comment]
On mobile the problem is even more prominent. For example when running the upload demo: https://nicegui.io/documentation/upload the user has to pick a file from the filesystem. That stops the execution of the webpage and when returning the connection can be lost. Resulting in a reload. Maybe we should look into ~~Web Workers or~~ Service Workers to keep the connection open?

**Reply by retsyo** [reply]
any progressing?

Or if we use pure Javascript to implement the whole page, is there still such problem?

**Reply by rodja** [reply]
>  if we use pure Javascript to implement the whole page, is there still such problem?

That would solve the problem. But you would need another framework than NiceGUI to do this. We have not yet tried to tackle the underlying issue with Service Workers.

**Comment by falkoschindler** [comment]
I just experimented with @rodja's code example and discussed the matter with ChatGPT. It seems like the whole JavaScript runtime (or parts of it) is paused while the print dialog is shown.

```py
label = ui.label()
ui.timer(0.1, lambda: label.set_text(f'{time.time():.3f}'))
ui.button('Print', on_click=lambda: ui.run_javascript('print()'))
```

**Comment by rodja** [comment]
I created feature request #2410 to find a suitable implementation. Please upvote!
